ludwig/decoders/sequence_decoders.py

```diff
@@ -13,18 +13,227 @@
 # limitations under the License.
 # ==============================================================================
 import logging
+from typing import Dict, Tuple
 
-from ludwig.constants import SEQUENCE, TEXT
+import torch
+import torch.nn as nn
+
+from ludwig.constants import HIDDEN, LOGITS, SEQUENCE, TEXT
 from ludwig.decoders.base import Decoder
 from ludwig.decoders.registry import register_decoder
+from ludwig.utils import strings_utils
 
 logger = logging.getLogger(__name__)
 
 
-# TODO(Justin): Implement sequence decoders.
+class RNNDecoder(nn.Module):
+    """GRU or RNN-based decoder."""
+
+    def __init__(self, hidden_size: int, vocab_size: int, cell_type: str):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.vocab_size = vocab_size
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.relu = nn.ReLU()
+        if cell_type == "gru":
+            self.rnn = nn.GRU(hidden_size, hidden_size, batch_first=True)
+        else:
+            self.rnn = nn.RNN(hidden_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, vocab_size)
```

Comment:
In the tf implementation I believe we had the option to have the same weights for embedding and out linear (being one the transpose of the other)

Reply:
Discussed offline. It looks like the weight sharing isn't present in version 0.4. I've added it back in this torch implementation, with a pointer to the Transformer paper.

Empirically, this gets us a lower loss on the twitterbots dataset, but it might be good to follow up with a PR to make this configurable.